src/Compilers/Core/Portable/FileSystem/PathUtilities.cs

```diff
@@ -773,6 +773,31 @@ public static bool IsValidFilePath([NotNullWhen(true)] string? fullPath)
         public static string NormalizeWithForwardSlash(string p)
             => DirectorySeparatorChar == '/' ? p : p.Replace(DirectorySeparatorChar, '/');
 
+        /// <summary>
+        /// Takes a normalized full path from <see cref="NormalizeWithForwardSlash(string)"/> and expands any '..' or '.'
+        /// into their equivalent representation.
+        /// </summary>
+        /// <returns>An equivalent path that does not contain any '..' or '.' path parts.</returns>
+        public static string ExpandNormalizedFullPathWithRelativeParts(string p)
+        {
+            Debug.Assert(p.StartsWith("/"));
+
+            var parts = GetPathParts(p);
+            Stack<string> resolvedParts = new Stack<string>();
```

Comment:
>Stack [](http://example.com/codeflow?start=12&length=5)

Consider using `ArrayBuilder` instead. (There are extension methods in `Microsoft.CodeAnalysis.ArrayBuilderExtensions` for `Push()` and `Pop()`.) That would allow using a pooled instance and wouldn't require `Reverse()` at the end.

Reply:
additional nit: consider using `var` for the type of `resolvedParts`.